File: new/top_k_frequent_elements_sorting.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // create a frequncy map
        // make an array out of the frequency map of frequncy, val. 
        // sort the array
        // get the last k elements
        unordered_map<int, int> frequency_map;
        for (const auto &num: nums) {
            frequency_map[num]++;
        }

        vector<pair<int, int>> arr;
        for (const auto &pair: frequency_map) {
            arr.push_back({pair.second, pair.first});
        }

        sort(arr.rbegin(), arr.rend());

        vector<int> result;
        for (int i = 0; i < k; i++) {
            result.push_back(arr[i].second);
        }

        return result;
    }   

    // time: O(nlogn) for sorting
    // space: O(n)
};

File: new/top_k_frequent_elements_sorting.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # using sorting
        # create a frequency map: num: count
        # make a list out of the frequency map [[frequency, num], ...] and sort it.
        # get the last k items's 2nd item from the list, it will have our top k freq element
        map = {}
        for num in nums:
            map[num] = 1 + map.get(num, 0)

        arr = []
        for key, value in map.items():  # use map.items to get the key value pairs
            arr.append([value, key])
        arr.sort()

        result = []
        while len(result) < k:
            result.append(arr.pop()[1])

        return result

        # time: O(nlogn) for sorting
        # space: O(n)
